src/Stratis.Bitcoin/Consensus/Validators/BlockValidator.cs

```diff
@@ -37,39 +48,117 @@ public interface IHeaderValidator
         }
 
         // <inheritdoc />
         public void VerifyBlockIntegrity(Block block, ChainedHeader chainedHeader)
         {
-            throw new System.NotImplementedException();
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeader), chainedHeader);
+
+            var validationContext = new ValidationContext { Block = block };
+
+            this.consensusRules.IntegrityValidationAsync(validationContext, null);
+
+            this.logger.LogTrace("(-)");
         }
     }
 
     // <inheritdoc />
     public class PartialValidation : IPartialValidation
     {
+        private readonly IConsensusRules consensusRules;
+        private readonly AsyncQueue<PartialValidationItem> asyncQueue;
+        private readonly ILogger logger;
+
+        public PartialValidation(IConsensusRules consensusRules, ILoggerFactory loggerFactory)
+        {
+            this.consensusRules = consensusRules;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName);
+
+            this.asyncQueue = new AsyncQueue<PartialValidationItem>(this.OnEnqueueAsync);
+        }
+
+        private async Task OnEnqueueAsync(PartialValidationItem item, CancellationToken cancellationtoken)
+        {
+            this.logger.LogTrace("({0}:'{1}')", nameof(item), item);
+
+            var validationContext = new ValidationContext {Block = item.ChainedHeaderBlock.Block};
+
+            await this.consensusRules.PartialValidationAsync(new ValidationContext { Block = item.ChainedHeaderBlock.Block, ChainedHeader = item.ChainedHeaderBlock.ChainedHeader }, null);
+
+            var partialValidationResult = new PartialValidationResult
+            {
+                ChainedHeaderBlock = item.ChainedHeaderBlock,
+                BanDurationSeconds = validationContext.BanDurationSeconds,
+                BanReason = validationContext.Error != null ? $"Invalid block received: {validationContext.Error.Message}" : string.Empty,
+                Succeeded = validationContext.Error != null
+            };
+
+            await item.PartialValidationCompletedAsyncCallback(partialValidationResult).ConfigureAwait(false);
+
+            this.logger.LogTrace("(-)");
+        }
+
         // <inheritdoc />
         public void StartPartialValidation(ChainedHeaderBlock chainedHeaderBlock, OnPartialValidationCompletedAsyncCallback onPartialValidationCompletedAsyncCallback)
         {
-            throw new System.NotImplementedException();
+            this.logger.LogTrace("({0}:'{1}')", nameof(chainedHeaderBlock), chainedHeaderBlock);
+
+            this.asyncQueue.Enqueue(new PartialValidationItem()
+            {
+                ChainedHeaderBlock = chainedHeaderBlock,
+                PartialValidationCompletedAsyncCallback = onPartialValidationCompletedAsyncCallback
+            });
+
+            this.logger.LogTrace("(-)");
+        }
+
+        public class PartialValidationItem
```

Comment:
This can be made private. Or completely removed in case you choose to initialize the validator with a callback. In that case we will have queue of blocks and won't need queueItem class